### PR TITLE
[ValueTracking] Improve `isImpliedCondCommonOperandWithConstants` to handle truncated LHS

### LINKED ISSUE #68514: Simplification for `(A ==/!= B) &/| (((cast)A) CMP C)`
Example:
https://godbolt.org/z/Pr8T7xsMz
```c
bool f(int a)
{
    bool t = a == 0;
    short t1 = a;
    bool t2 = t1 >= 3;
    return t & t2;
}

bool g(int b)
{
    bool t = b == 3;
    short t1 = b;
    bool t2 = t1 >= 3;
    return t | t2;
}
```

The example above is taken from the following:
https://github.com/gcc-mirror/gcc/blob/master/gcc/testsuite/gcc.dg/tree-ssa/cmpbit-6.c
https://github.com/gcc-mirror/gcc/blob/master/gcc/testsuite/gcc.dg/tree-ssa/cmpbit-7.c

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch improves `isImpliedCondCommonOperandWithConstants` to handle patterns like `L0 == trunc(R0)` or `R0 == trunc(L0)`. I am sure that we don't need to handle `sext/zext` cases since `icmp pred sext/zext(X), C` will be simplified into `icmp pred X, C1`.

Alive2: https://alive2.llvm.org/ce/z/FLcm2h

Fixes #68514.

### CODE REVIEW

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -8231,14 +8231,30 @@ isImpliedCondMatchingOperands(CmpInst::Predicate LPred,
...
   ConstantRange DomCR = ConstantRange::makeExactICmpRegion(LPred, LC);
   ConstantRange CR = ConstantRange::makeExactICmpRegion(RPred, RC);
+
+  if (L0 == R0)
+    ; // noop
+  // Example: icmp eq X, 3 --> icmp sgt trunc(X), 2
+  else if (match(R0, m_Trunc(m_Specific(L0))))
+    DomCR = DomCR.truncate(RC.getBitWidth());
+  // Example: icmp slt trunc(X), 3 --> icmp ne X, 3
+  else if (match(L0, m_Trunc(m_Specific(R0)))) {
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:8249]: Can you also check `zext`?

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 2]
```diff
@@ -8272,8 +8288,19 @@ static std::optional<bool> isImpliedCondICmps(const ICmpInst *LHS,
...
   // constants (not necessarily matching)?
   const APInt *LC, *RC;
-  if (L0 == R0 && match(L1, m_APInt(LC)) && match(R1, m_APInt(RC)))
-    return isImpliedCondCommonOperandWithConstants(LPred, *LC, RPred, *RC);
+  if (match(L1, m_APInt(LC)) && match(R1, m_APInt(RC))) {
+    if (auto Res = isImpliedCondCommonOperandWithConstants(L0, LPred, *LC, R0,
+                                                           RPred, *RC))
+      return Res;
+
+    if (match(L0, m_Trunc(m_Specific(R0)))) {
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:8296]: Isn't this covered by the `match(L0, m_Trunc(R0))...` above?

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:8296]: No. We need the law of excluded middle to cover more cases.
Example:
```
define i1 @g(i32 %a) {
; CHECK-LABEL: define i1 @g(
; CHECK-SAME: i32 [[A:%.*]]) {
; CHECK-NEXT:  entry:
; CHECK-NEXT:    [[B:%.*]] = trunc i32 [[A]] to i16
; CHECK-NEXT:    [[CMP2:%.*]] = icmp sgt i16 [[B]], 2
; CHECK-NEXT:    ret i1 [[CMP2]]
;
entry:
  %cmp = icmp eq i32 %a, 3
  %b = trunc i32 %a to i16
  %cmp2 = icmp sgt i16 %b, 2
  %or9 = or i1 %cmp, %cmp2
  ret i1 %or9
}
```

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 1 above)
Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:8249]: `icmp pred zext/sext(X), C` will be simplified to `icmp pred X, C1`. So we don't need to handle these cases.

**Comment by dtcxzyw** [comment]
Ping.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 3]
```diff
@@ -8231,14 +8231,30 @@ isImpliedCondMatchingOperands(CmpInst::Predicate LPred,
...
+    ; // noop
+  // Example: icmp eq X, 3 --> icmp sgt trunc(X), 2
+  else if (match(R0, m_Trunc(m_Specific(L0))))
+    DomCR = DomCR.truncate(RC.getBitWidth());
+  // Example: icmp slt trunc(X), 3 --> icmp ne X, 3
+  else if (match(L0, m_Trunc(m_Specific(R0)))) {
+    // Try to prove by negation
+    DomCR = DomCR.inverse();
+    CR = CR.inverse();
+    std::swap(DomCR, CR);
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:8253]: Is there any reason for this swap?

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:8253]: `A -> B` <-> `¬B -> ¬A`

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:8253]: err I mean following logic doesn't seem to care about `DomCR` vs `CR`.
But guess swap is clearer.

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:8253]: `ConstantRange::difference` is non-commutative. BTW, the bit width of `DomCR` and `CR` are not equal.

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:8253]: oof yeah.

**Comment by dtcxzyw** [comment]
Closed as the original issue has been fixed by https://github.com/llvm/llvm-project/pull/70335.

**Pull request closed without merging**